engine/src/main/java/org/hibernate/search/engine/common/impl/IndexManagerBuildingStateHolder.java

```diff
@@ -35,31 +38,39 @@
  */
 class IndexManagerBuildingStateHolder {
 
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
 	private static final OptionalConfigurationProperty<String> INDEX_BACKEND_NAME =
 			ConfigurationProperty.forKey( "backend" ).asString().build();
 
 	private static final OptionalConfigurationProperty<BeanReference<? extends BackendFactory>> BACKEND_TYPE =
 			ConfigurationProperty.forKey( "type" ).asBeanReference( BackendFactory.class ).build();
 
-	private final RootBuildContext rootBuildContext;
+	private final BeanProvider beanProvider;
 	private final ConfigurationPropertySource propertySource;
+	private final RootBuildContext rootBuildContext;
 	private final ConfigurationPropertySource defaultIndexPropertySource;
 
 	private final Map<String, BackendBuildingState<?>> backendBuildingStateByName = new HashMap<>();
 	private final Map<String, IndexManagerBuildingStateImpl<?>> indexManagerBuildingStateByName = new HashMap<>();
 
-	IndexManagerBuildingStateHolder(RootBuildContext rootBuildContext,
-			ConfigurationPropertySource propertySource) {
-		this.rootBuildContext = rootBuildContext;
+	IndexManagerBuildingStateHolder(BeanProvider beanProvider, ConfigurationPropertySource propertySource,
+			RootBuildContext rootBuildContext) {
+		this.beanProvider = beanProvider;
 		this.propertySource = propertySource;
+		this.rootBuildContext = rootBuildContext;
 		this.defaultIndexPropertySource = propertySource.withMask( "indexes.default" );
 	}
 
 	public IndexManagerBuildingState<?> startBuilding(String indexName, boolean multiTenancyEnabled) {
 		ConfigurationPropertySource indexPropertySource = propertySource.withMask( "indexes." + indexName )
 				.withFallback( defaultIndexPropertySource );
-		// TODO HSEARCH-3442 more checks on the backend name (is non-null, non-empty) => use INDEX_BACKEND_NAME.getOrThrow?
-		String backendName = INDEX_BACKEND_NAME.get( indexPropertySource ).get();
+		String backendName = INDEX_BACKEND_NAME.getOrThrow(
+				indexPropertySource,
+				key -> log.indexBackendCannotBeNullOrEmpty(
+						indexName, key, INDEX_BACKEND_NAME.resolveOrRaw( defaultIndexPropertySource )
```

Comment:
Instead of use again `INDEX_BACKEND_NAME' inside the function as a closure variable, you could think use a bi-function such as: 
>  ( key, resolved ) ->  log.indexBackendCannotBeNullOrEmpty( indexName. key, resolved )

to gather the resolved value.

Reply:
That will not work here.
To be clear, the key being passed to the lambda has already been resolved. Here, I add an explicit call to resolve because I want to resolve the key against a different configuration source: the default one.
Basically `key` here is resolved to something like `hibernate.search.indexes.myIndex.backend`, whereas `INDEX_BACKEND_NAME.resolveOrRaw( defaultIndexPropertySource )` returns `hibernate.search.indexes.default.backend`.
It's definitely not clear, though. I will add a comment.

Reply:
Makes sense. Sorry. I didn't notice the use of two different property sources.